packages/daemon/DESIGN.md

```diff
@@ -0,0 +1,215 @@
+
+A Promise represents a result from either the past or the future.
+Eventual Send displaces a result in both space and time.
+That is, a program can use eventual send to uniformly interact with a result
+regardless of whether the eventual result is local or remote, pending or
+settled.
+Furthermore, Eventual Send can pipeline[^pipeline] messages to an eventual result,
+like invoking a method on the result of a prior method invocation, to avoid
+unnecessary round trips through the underlying CapTP connection.
+
+[^pipeline]: See [Mark's explanation of pipelines][pipeline-erights], the
+    [Cap'n Proto explanation][pipeline-capnproto], [Wikipedia][pipeline-wikipedia], or
+    the [ECMAScript standard proposal][pipeline-proposal].
+
+[pipeline-erights]: http://erights.org/elib/distrib/pipeline.html
+[pipeline-capnproto]: https://capnproto.org/rpc.html
+[pipeline-wikipedia]: https://en.wikipedia.org/wiki/Futures_and_promises#Promise_pipelining
+[pipeline-proposal]: https://www.proposals.es/proposals/Support%20for%20Distributed%20Promise%20Pipelining
+
+Much hinges on what happens when a program restarts.
+Like most processes, Endo workers start without any memory from their previous
+run.
+So, Endo programs must manually persist[^persistence] anything they wish to
+remember after a restart.
+
+[^persistence]: Aside: Manual persistence is as opposed to Orthogonal
+    Persistence, where a program resumes exactly where it left off when it
+    restarts, albeit it by replaying every message it recieved since the last
+    snapshot or inception.  The architecture of Endo allows for Orthogonal
+    Persistence to be built on top of a manually persisted worker and a durable
+    reference to external storage for snapshots, message log, and virtual
+    objects.
+
+Endo applications ultimately receive all their data and powers from the Endo
+daemon with the user's consent.
+Endo persists **data** between restarts with a content address store in the
+user's home directory.
+Endo persists **powers** between restarts by storing formulas in the user's
+home directory.
+A formula describes how Endo should recreate a value, including powerful
+capability bearing objects if it restarts.
+
+When Endo runs a confined program, it provides a Powers object that the
+confined program can use to **request** powers from the user.
+The Powers object is entangled with a corresponding Inbox.
+The user interacts with the Inbox to respond to requests from the Powers
+holder.
+
+Both the Inbox held by the user (host) and the Powers held by program (guest)
+refer to formulas with pet names, and have separate pet name spaces.
+Changing a pet name does not alter or delete the underlying formula, but a
+formula that is unreachable from the user's pet names may be garbage collected.
+
+- The program can use the Powers object to request an ephemeral value
+  from the user.
+  The Powers will not retain the resulting formula.
+
+  ```js
+  const file = E(powers).request('a file');
```

Comment:
As I read this, I don't know what `'a file'` is an example of, or what a realistic form of this example might be. A bit of such suspense is fine, because you cannot explain everything before everything else. But as I read ahead, this was not resolved within my "what does this really mean?" horizon. What does this mean? When the program makes a request like this, what is it actually asking for? What does the user see?

Reply:
```suggestion
  const fileP = E(powers).request('a file');
```
to make it clear that the expression returns a promise?

or possibly
```suggestion
  const file = await E(powers).request('a file');
```
though I prefer the first form over this ;)